### PR TITLE
Add support for neo4j vector index

### PR DESCRIPTION
**Pull request by tomasonjo**
Neo4j has a vector index that can be used for vector retrieval in RAG in other applications. It can be expanded with retrieval query, so that it can traverse through the graph after the initial vector search if needed.

### CODE REVIEW

File: docs/api/retrieval_model_clients/Neo4jRM.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    k: int = 5,
+    retrieval_query: str = None,
+    embedding_provider: str = "openai",
+    embedding_model: str = "text-embedding-ada-002",
+)
+```
+
+**Environment Variables:**
+
+You need to define the credentials as environment variables:
```
Comment (jexp) [inline docs/api/retrieval_model_clients/Neo4jRM.md:21]: add OPENAI_API_KEY as required env variable in docs + example

File: dspy/retrieve/neo4j_rm.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        contents = []
+        retrieval_query = self.retrieval_query or f"RETURN node.{self.text_node_property} AS text"
+        for vector in query_vectors:
+            records, _, _ = self.driver.execute_query(
+                DEFAULT_INDEX_QUERY + retrieval_query,
+                {"embedding": vector, "index": self.index_name, "k": k or self.k},
+                database_=self.database,
+            )
+            contents.extend([dotdict({"long_text": r["text"]}) for r in records])
+        return contents
```
Comment (jexp) [inline dspy/retrieve/neo4j_rm.py:107]: sort by score and take top-k results

Comment (tomasonjo) [reply dspy/retrieve/neo4j_rm.py]: Fixed

**Comment by arnavsinghvi11** [comment]
Thanks for this addition @tomasonjo @jexp !

**Pull request merged**